src/Npgsql/Replication/Logical/Protocol/NpgsqlLogicalReplicationConnectionExtensions.cs

```diff
@@ -0,0 +1,352 @@
+        /// for more information.
+        /// </remarks>
+        /// <param name="connection">The <see cref="NpgsqlLogicalReplicationConnection"/> to use for creating the
+        /// replication slot</param>
+        /// <param name="slotName">The name of the slot to create. Must be a valid replication slot name (see
+        /// <a href="https://www.postgresql.org/docs/current/warm-standby.html#STREAMING-REPLICATION-SLOTS-MANIPULATION">https://www.postgresql.org/docs/current/warm-standby.html#STREAMING-REPLICATION-SLOTS-MANIPULATION</a>).
+        /// </param>
+        /// <param name="temporarySlot">
+        /// <see langword="true"/> if this replication slot shall be temporary one; otherwise <see langword="false"/>.
+        /// Temporary slots are not saved to disk and are automatically dropped on error or when the session has finished.
+        /// </param>
+        /// <param name="slotSnapshotInitMode">
+        /// A <see cref="SlotSnapshotInitMode"/> to specify what to do with the snapshot created during logical slot
+        /// initialization. <see cref="SlotSnapshotInitMode.Export"/>, which is also the default, will export the
+        /// snapshot for use in other sessions. This option can't be used inside a transaction.
+        /// <see cref="SlotSnapshotInitMode.Use"/> will use the snapshot for the current transaction executing the
+        /// command. This option must be used in a transaction, and <see cref="SlotSnapshotInitMode.Use"/> must be the
+        /// first command run in that transaction. Finally, <see cref="SlotSnapshotInitMode.NoExport"/> will just use
+        /// the snapshot for logical decoding as normal but won't do anything else with it.
+        /// </param>
+        /// <param name="cancellationToken">
+        /// The token to monitor for cancellation requests.
+        /// The default value is <see cref="CancellationToken.None"/>.
+        /// </param>
+        /// <returns>
+        /// A <see cref="NpgsqlPgOutputReplicationSlot"/> that wraps the newly-created replication slot.
+        /// </returns>
+        [PublicAPI]
+        public static async Task<NpgsqlPgOutputReplicationSlot> CreateReplicationSlot(
+            this NpgsqlLogicalReplicationConnection connection,
+            string slotName,
+            bool temporarySlot = false,
+            SlotSnapshotInitMode? slotSnapshotInitMode = null,
+            CancellationToken cancellationToken = default)
+        {
+            // We don't enter NoSynchronizationContextScope here since we (have to) do it in CreateReplicationSlotForPlugin, because
+            // otherwise it couldn't be set for external plugins.
+            var options =
+                await connection.CreateReplicationSlotForPlugin(slotName, "pgoutput", temporarySlot, slotSnapshotInitMode, cancellationToken);
```

Comment:
> Is this class really about the pgoutput output plugin (and so should be named that way)?

It is (kind of) see https://github.com/npgsql/npgsql/pull/3281#discussion_r515520691

> I think that means users are expected to select their output plugin by importing the right namespace, which seems a bit problematic...

😢 This was intended to be a feature (kind of per-plugin overloads of the same method).

> but do we really need to that (i.e. does the slot type need to distinguish between the plugins)?

Yes. The slot and the plugin are coupled. You can't use a slot created for pgoutput to stream test_decoding or the other way around.

Reply:
> This was intended to be a feature (kind of per-plugin overloads of the same method).

OK. The main problem I see with this idea is discoverability - imagine a new user who has no idea that replication even has plugins, and has no idea about the namespaces; the only way they can ever discover them is probably by looking in the docs. One of the very cool aspects of extension methods, is that plugin can add them to your default namespace, and they just "light up" with Intellisense; EF Core does this very nicely: you just add a reference to the Npgsql provider, and UseNpgsql just appears (because it's defined under Microsoft.EntityFrameworkCore).

Omitting the plugin name from the method also makes it unclear which plugin is being used - instead of the API describing itself, you have to go and see which namespaces are being imported (and to know you have to do that). It also makes it very cumbersome to ever use two plugins in the same source file (I agree this should be rare with replication but...).

Anyway, these are my arguments... :)

>> but do we really need to that (i.e. does the slot type need to distinguish between the plugins)?
>
> Yes. The slot and the plugin are coupled. You can't use a slot created for pgoutput to stream test_decoding or the other way around.

OK, you're right - this makes sense since we need to have different versions of StartReplication, since their return type is completely different.

This makes me want to move the plugin-specific StartReplication methods to the slots themselves. This works really nicely when you create the slot from the connection (the slot would be bound to the connection), but I'm not sure what it would look like if the slot already exists. How do you feel about this?

Reply:
> This makes me want to move the plugin-specific StartReplication methods to the slots themselves. This works really nicely when you create the slot from the connection (the slot would be bound to the connection), but I'm not sure what it would look like if the slot already exists. How do you feel about this?

🤣 I've already been there.
It works.
https://github.com/Brar/npgsql/blob/LogicalReplicationAPI/src/Npgsql/Replication/Logical/Protocol/NpgsqlPgOutputReplicationSlot.cs

I dropped this Idea because I felt that it deviates too far from the PostgreSQL streaming replication protocol where you start replication on the connection using a slot and not on the slot itself, which is especially true for physical replication where you can even start replication without a slot.

The other side is that indeed you rather replicate the information that is stored in the slot, so it feels really natural to start replication on the slot (which you somehow have to bind to a connection to be able to start replication).

Reply:
> Bottom line, I can see us going either way but am leaning slightly towards keeping things as they are now.

That's exactly how I ended up with the current API. It feels good to have you follow along and confirm my thought process.

Reply:
https://github.com/Brar/npgsql/pull/5